### PR TITLE
add shrink-mds playbook

### PR DESCRIPTION
**Pull request by rishabh-d-dave**
Add a playbook that removes a MDS node from an deployed cluster.

### CODE REVIEW

**Comment by rishabh-d-dave** [comment]
@guits @dsavineau 

I've got few questions -

* `shrink-mon` and `shrink-osd` tox scenarios don't run tests after they shrink the cluster but I suppose we should run tests. What's your opinion?

* shrink-mon.yml aborts when the MON to be deleted is last/only in cluster. Should `shrink-mds.yml` behave in the same way? I personally think we should allow removing only MDS from the cluster since MDS is not critical to the cluster. Someone just might be removing CephFS from the cluster.

  Also, I think we should add allowing only monitor from cluster. Maybe we make the playbook demanding an extra flag `remove_only_MON_too`.

* What's the best way to stop MDS? `systemctl stop ceph-mds@{hostname}`, `ceph tell mds_to_kill exit` or `ceph mds fail mds_to_kill`

**Comment by rishabh-d-dave** [comment]
About last push -
  * modified shrink-mds.yml so that it works for containers too (now `dev-centos-containerized-shirnk_mds` is passing too).
  * minor corrections in shrink-mds.yml

**Comment by rishabh-d-dave** [comment]
jenkins test dev-centos-non_container-shirnk_mds

**Comment by rishabh-d-dave** [comment]
jenkins test dev-centos-container-shirnk_mds

**Comment by rishabh-d-dave** [comment]
jenkins test dev-centos-non_container-shrink_mds

**Comment by rishabh-d-dave** [comment]
jenkins test dev-centos-container-shrink_mds

**Comment by rishabh-d-dave** [comment]
@guits @dsavineau can you please review this PR?

**Comment by rishabh-d-dave** [comment]
jenkins test dev-centos-non_container-shrink_mds

**Comment by rishabh-d-dave** [comment]
jenkins test dev-centos-container-shrink_mds

File: infrastructure-playbooks/shrink-mds.yml [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+          {{ container_binary }} exec ceph-mon-{{ hostvars[groups
+          [mon_group_name][0]]['ansible_hostname'] }}
+      when: containerized_deployment | bool
+
+    - name: exit playbook, if can not connect to the cluster
+      command: >
+        {{ container_exec_cmd | default('') }} timeout 5 ceph --cluster
+        {{ cluster }} health
+      register: ceph_health
+      until: ceph_health.stdout.find("HEALTH") > -1
```
Comment (guits) [inline infrastructure-playbooks/shrink-mds.yml]:
```suggestion
      until: ceph_health.rc
```

Comment (rishabh-d-dave) [reply infrastructure-playbooks/shrink-mds.yml]: Done.

File: infrastructure-playbooks/shrink-mds.yml [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+# ansible-playbook -e ireallymeanit=yes|no shrink-mds.yml
+#     Overrides the prompt using -e option. Can be used in
+#     automation scripts to avoid interactive prompt.
+- name: gather facts and check the init system
+  hosts:
+    - "{{ mon_group_name | default('mons') }}"
+    - "{{ mds_group_name | default('mdss') }}"
+  become: true
+  tasks:
+    - debug: msg="gather facts on all Ceph hosts for following reference"
```
Comment (guits) [inline infrastructure-playbooks/shrink-mds.yml]: you must import the role ceph-defaults here
```suggestion
    - import_role:
        name: ceph-defaults
    - debug: msg="gather facts on all Ceph hosts for following reference"
```

File: infrastructure-playbooks/shrink-mds.yml [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+- name: perform checks, remove mds and print cluster health
+  hosts: localhost
+  become: true
+  vars_prompt:
+    - name: ireallymeanit
+      prompt: Are you sure you want to shrink the cluster?
+      default: 'no'
+      private: no
+  vars:
+    mds_group_name: mdss
```
Comment (guits) [inline infrastructure-playbooks/shrink-mds.yml]: is it really needed?

File: infrastructure-playbooks/shrink-mds.yml [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+
+    - name: exit playbook, if user did not mean to shrink cluster
+      when: ireallymeanit != 'yes'
+      fail:
+        msg: "Exiting shrink-mds playbook, no mds was removed.
+           To shrink the cluster, either say 'yes' on the prompt or
+           or use `-e ireallymeanit=yes` on the command line when
+           invoking the playbook"
+
+    - import_role:
```
Comment (guits) [inline infrastructure-playbooks/shrink-mds.yml]: you should move this task before any other task here

Comment (rishabh-d-dave) [reply infrastructure-playbooks/shrink-mds.yml]: Done.

File: infrastructure-playbooks/shrink-mds.yml (hunk 3 above)
Comment (rishabh-d-dave) [reply infrastructure-playbooks/shrink-mds.yml]: Um.. I guess. I've removed this too.

File: infrastructure-playbooks/shrink-mds.yml (hunk 2 above)
Comment (rishabh-d-dave) [reply infrastructure-playbooks/shrink-mds.yml]: Should `ceph-facts` too move here?

**Comment by rishabh-d-dave** [comment]
A TODO from the patch got lost during updates...

@guits @dsavineau 
should we run Python tests at the end of `shrink_mds`?

File: infrastructure-playbooks/shrink-mds.yml [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+          {{ container_binary }} exec ceph-mon-{{ hostvars[groups
+          [mon_group_name][0]]['ansible_hostname'] }}
+      when: containerized_deployment | bool
+
+    - name: exit playbook, if can not connect to the cluster
+      command: >
+        {{ container_exec_cmd | default('') }} timeout 5 ceph --cluster
+        {{ cluster }} health
+      register: ceph_health
+      until: ceph_health.rc
```
Comment (guits) [inline infrastructure-playbooks/shrink-mds.yml]:
```suggestion
      until: ceph_health is succeeded
```

File: infrastructure-playbooks/shrink-mds.yml [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+      failed_when: ceph_health.rc != 0
```
Comment (guits) [inline infrastructure-playbooks/shrink-mds.yml]:
```suggestion
```

File: infrastructure-playbooks/shrink-mds.yml (hunk 5 above)
Comment (rishabh-d-dave) [reply infrastructure-playbooks/shrink-mds.yml]: Done.

File: infrastructure-playbooks/shrink-mds.yml (hunk 6 above)
Comment (rishabh-d-dave) [reply infrastructure-playbooks/shrink-mds.yml]: done.

**Comment by rishabh-d-dave** [comment]
jenkins test dev-centos-non_container-shrink_mds

**Comment by rishabh-d-dave** [comment]
jenkins test dev-centos-container-shrink_rgw

**Comment by rishabh-d-dave** [comment]
jenkins test dev-centos-container-shrink_mds

File: infrastructure-playbooks/shrink-mds.yml (hunk 2 above)
Comment (guits) [reply infrastructure-playbooks/shrink-mds.yml]: not needed

File: infrastructure-playbooks/shrink-mds.yml [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+  vars_prompt:
+    - name: ireallymeanit
+      prompt: Are you sure you want to shrink the cluster?
+      default: 'no'
+      private: no
+  pre_tasks:
+    - import_role:
+        name: ceph-defaults
+
+    - import_role:
```
Comment (guits) [inline infrastructure-playbooks/shrink-mds.yml]: move this task after the `exit playbook, if user did not mean to shrink cluster`

we shouldn't have to wait for ceph-facts to be run to potentially fail in one of the 3 following task

**Comment by guits** [comment]
jenkins test dev-centos-non_container-shrink_mds

**Comment by guits** [comment]
jenkins test dev-centos-container-shrink_mds

**Comment by guits** [comment]
jenkins test dev-centos-non_container-shrink_mds

**Comment by guits** [comment]
jenkins test dev-centos-container-shrink_mds

**Comment by guits** [comment]
jenkins test dev-centos-container-shrink_mds

**Comment by guits** [comment]
jenkins test dev-centos-non_container-shrink_mds

**Comment by guits** [comment]
jenkins test dev-centos-non_container-shrink_mds

**Comment by guits** [comment]
jenkins test dev-centos-container-shrink_mds

**Comment by guits** [comment]
jenkins test dev-centos-non_container-shrink_mds

**Comment by rishabh-d-dave** [comment]
jenkins test dev-centos-non_container-shrink_mds

**Comment by rishabh-d-dave** [comment]
jenkins test dev-centos-container-shrink_mds

**Comment by rishabh-d-dave** [comment]
jenkins test dev-centos-container-shrink_mds

**Comment by rishabh-d-dave** [comment]
jenkins test dev-centos-non_container-shrink_mds

**Comment by rishabh-d-dave** [comment]
jenkins test dev-centos-container-shrink_mds

**Comment by rishabh-d-dave** [comment]
jenkins test dev-centos-non_container-shrink_mds

**Comment by rishabh-d-dave** [comment]
jenkins test dev-centos-container-shrink_mds

File: infrastructure-playbooks/shrink-mds.yml [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+
+        - name: fail if mds just killed is being reported as active or standby
+          fail:
+            msg: "mds node {{ mds_to_kill }} still up and running."
+          when:
+            - (mds_to_kill in active_mdss | default([])) or
+              (mds_to_kill in standby_mdss | default([]))
+
+    - name: delete the filesystem too if deleted the last mds too
+      command: >
```
Comment (guits) [inline infrastructure-playbooks/shrink-mds.yml]:
```suggestion
      command: "{{ container_exec_cmd | default('') }} ceph --cluster {{ cluster }} fs rm --yes-i-really-mean-it {{ cephfs }}"
```

File: infrastructure-playbooks/shrink-mds.yml [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+        {{ container_exec_cmd | default('') }} ceph --cluster {{ cluster }}
```
Comment (guits) [inline infrastructure-playbooks/shrink-mds.yml]:
```suggestion
```

File: infrastructure-playbooks/shrink-mds.yml [hunk 10]
```diff
@@ -0,0 +1,147 @@
...
+        fs rm --yes-i-really-mean-it {{ cephfs }}
```
Comment (guits) [inline infrastructure-playbooks/shrink-mds.yml]:
```suggestion
```

**Comment by guits** [comment]
jenkins test dev-centos-container-shrink_mds

**Comment by guits** [comment]
jenkins test dev-centos-non_container-shrink_mds

**Pull request merged**